src/libraries/System.Text.Json/src/System/Text/Json/Serialization/JsonConverterOfT.cs

```diff
@@ -331,11 +334,9 @@ internal bool TryWrite(Utf8JsonWriter writer, in T value, JsonSerializerOptions
 
                 // For boxed reference types: do not push when boxed in order to avoid false positives
                 //   when we run the ContainsReferenceForCycleDetection check for the converter of the unboxed value.
-                if (!CanBePolymorphic)
-                {
-                    resolver.PushReferenceForCycleDetection(value);
-                    ignoreCyclesPopReference = true;
-                }
+                Debug.Assert(!CanBePolymorphic);
```

Comment:
Why does this hold? IIRC `JsonConverter<object>` can be polymorphic and has `ConverterStrategy.Object`.